`-Wdeprecated-copy` warnings generated by the `test::cxx11_allocator`

Compiling the test suite using, for example:
```console
./b2 -a libs/unordered/test cxxstd=20 toolset=clang-12
```
on Ubuntu 20.04 will generate many such warnings:
```console
In file included from libs/unordered/test/unordered/assign_tests.cpp:15:
libs/unordered/test/unordered/../objects/cxx11_allocator.hpp:265:5: warning: definition of implicit copy assignment operator for 'cxx11_allocator<boost::unordered::detail::ptr_node<std::pair<const test::object, test::object>>, test::propagate_assign>' is deprecated because it has a user-declared copy constructor [-Wdeprecated-copy]
    cxx11_allocator(cxx11_allocator const& x) : cxx11_allocator_base<T>(x) {}
    ^
./boost/unordered/detail/implementation.hpp:452:20: note: in implicit copy assignment operator for 'test::cxx11_allocator<boost::unordered::detail::ptr_node<std::pair<const test::object, test::object>>, test::propagate_assign>' first required here
          second() = x.second();
                   ^
./boost/unordered/detail/implementation.hpp:3462:25: note: in instantiation of member function 'boost::unordered::detail::compressed<test::cxx11_allocator<boost::unordered::detail::ptr_bucket, test::propagate_assign>, test::cxx11_allocator<boost::unordered::detail::ptr_node<std::pair<const test::object, test::object>>, test::propagate_assign>>::assign' requested here
            allocators_.assign(x.allocators_);
                        ^
./boost/unordered/detail/implementation.hpp:3425:13: note: in instantiation of function template specialization 'boost::unordered::detail::table<boost::unordered::detail::map<test::cxx11_allocator<test::object, test::propagate_assign>, test::object, test::object, test::hash, test::equal_to>>::assign<std::integral_constant<bool, true>>' requested here
            assign(x, is_unique,
            ^
./boost/unordered/unordered_map.hpp:166:16: note: in instantiation of function template specialization 'boost::unordered::detail::table<boost::unordered::detail::map<test::cxx11_allocator<test::object, test::propagate_assign>, test::object, test::object, test::hash, test::equal_to>>::assign<std::integral_constant<bool, true>>' requested here
        table_.assign(x.table_, boost::unordered::detail::true_type());
               ^
libs/unordered/test/unordered/assign_tests.cpp:38:9: note: in instantiation of member function 'boost::unordered::unordered_map<test::object, test::object, test::hash, test::equal_to, test::cxx11_allocator<test::object, test::propagate_assign>>::operator=' requested here
      x = x;
        ^
libs/unordered/test/unordered/assign_tests.cpp:251:18: note: in instantiation of function template specialization 'assign_tests::assign_tests1<boost::unordered::unordered_map<test::object, test::object, test::hash, test::equal_to, test::cxx11_allocator<test::object, test::propagate_assign>>>' requested here
  UNORDERED_TEST(assign_tests1,
                 ^
```

This also shows up in the CI:
https://github.com/boostorg/unordered/runs/4216196265?check_suite_focus=true#step:6:847

--- a/test/objects/cxx11_allocator.hpp
+++ b/test/objects/cxx11_allocator.hpp
@@ -13,7 +13,14 @@
 #include "../helpers/fwd.hpp"
 #include "../helpers/memory.hpp"
 
-namespace test {
+#if defined(BOOST_NO_CXX11_NOEXCEPT)
+#define BOOST_UNORDERED_NOEXCEPT
+#else
+#define BOOST_UNORDERED_NOEXCEPT noexcept
+#endif
+
+namespace test
+{
   struct allocator_false
   {
     enum
@@ -178,7 +185,36 @@ namespace test {
       detail::tracker.allocator_ref();
     }
 
+#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+    cxx11_allocator_base(cxx11_allocator_base&& x) BOOST_UNORDERED_NOEXCEPT
+    {
+      tag_ = x.tag_;
+      selected_ = x.selected_;
+
+      x.tag_ = -1;
+      x.selected_ = -1;
+    }
+#endif
+
     ~cxx11_allocator_base() { detail::tracker.allocator_unref(); }
+
+    cxx11_allocator_base& operator=(cxx11_allocator_base const& x)
+    {
+      tag_ = x.tag_;
+      selected_ = x.selected_;
+      return *this;
+    }
+
+#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+    cxx11_allocator_base& operator=(
+      cxx11_allocator_base&& x) BOOST_UNORDERED_NOEXCEPT
+    {
+      tag_ = x.tag_;
+      selected_ = x.selected_;
+
+      return *this;
+    }
+#endif
 
     pointer address(reference r) { return pointer(&r); }
 
@@ -264,6 +300,20 @@ namespace test {
 
     cxx11_allocator(cxx11_allocator const& x) : cxx11_allocator_base<T>(x) {}
 
+    cxx11_allocator& operator=(cxx11_allocator const& x)
+    {
+      cxx11_allocator_base<T>::operator=(x);
+      return *this;
+    }
+
+#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_UNORDERED_NOEXCEPT
+    {
+      cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
+      return *this;
+    }
+#endif
+
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.
     bool operator==(cxx11_allocator const& x) const
@@ -306,6 +356,20 @@ namespace test {
     }
 
     cxx11_allocator(cxx11_allocator const& x) : cxx11_allocator_base<T>(x) {}
+
+    cxx11_allocator& operator=(cxx11_allocator const& x)
+    {
+      cxx11_allocator_base<T>::operator=(x);
+      return *this;
+    }
+
+#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
+    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_UNORDERED_NOEXCEPT
+    {
+      cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
+      return *this;
+    }
+#endif
 
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.

--- a/test/objects/cxx11_allocator.hpp
+++ b/test/objects/cxx11_allocator.hpp
@@ -181,12 +181,9 @@ namespace test
 
     ~cxx11_allocator_base() { detail::tracker.allocator_unref(); }
 
-    cxx11_allocator_base& operator=(cxx11_allocator_base const& x)
-    {
-      tag_ = x.tag_;
-      selected_ = x.selected_;
-      return *this;
-    }
+#if !defined(BOOST_NO_CXX11_DEFAULTED_FUNCTIONS)
+    cxx11_allocator_base& operator=(cxx11_allocator_base const& x) = default;
+#endif
 
     pointer address(reference r) { return pointer(&r); }
 
@@ -272,11 +269,9 @@ namespace test
 
     cxx11_allocator(cxx11_allocator const& x) : cxx11_allocator_base<T>(x) {}
 
-    cxx11_allocator& operator=(cxx11_allocator const& x)
-    {
-      cxx11_allocator_base<T>::operator=(x);
-      return *this;
-    }
+#if !defined(BOOST_NO_CXX11_DEFAULTED_FUNCTIONS)
+    cxx11_allocator& operator=(cxx11_allocator const& x) = default;
+#endif 
 
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.
@@ -321,11 +316,9 @@ namespace test
 
     cxx11_allocator(cxx11_allocator const& x) : cxx11_allocator_base<T>(x) {}
 
-    cxx11_allocator& operator=(cxx11_allocator const& x)
-    {
-      cxx11_allocator_base<T>::operator=(x);
-      return *this;
-    }
+#if !defined(BOOST_NO_CXX11_DEFAULTED_FUNCTIONS)
+    cxx11_allocator& operator=(cxx11_allocator const& x) = default;
+#endif
 
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.

--- a/test/objects/cxx11_allocator.hpp
+++ b/test/objects/cxx11_allocator.hpp
@@ -13,12 +13,6 @@
 #include "../helpers/fwd.hpp"
 #include "../helpers/memory.hpp"
 
-#if defined(BOOST_NO_CXX11_NOEXCEPT)
-#define BOOST_UNORDERED_NOEXCEPT
-#else
-#define BOOST_UNORDERED_NOEXCEPT noexcept
-#endif
-
 namespace test
 {
   struct allocator_false
@@ -186,7 +180,7 @@ namespace test
     }
 
 #if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator_base(cxx11_allocator_base&& x) BOOST_UNORDERED_NOEXCEPT
+    cxx11_allocator_base(cxx11_allocator_base&& x) BOOST_NOEXCEPT
     {
       tag_ = x.tag_;
       selected_ = x.selected_;
@@ -207,7 +201,7 @@ namespace test
 
 #if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
     cxx11_allocator_base& operator=(
-      cxx11_allocator_base&& x) BOOST_UNORDERED_NOEXCEPT
+      cxx11_allocator_base&& x) BOOST_NOEXCEPT
     {
       tag_ = x.tag_;
       selected_ = x.selected_;
@@ -307,7 +301,7 @@ namespace test
     }
 
 #if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_UNORDERED_NOEXCEPT
+    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_NOEXCEPT
     {
       cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
       return *this;
@@ -364,7 +358,7 @@ namespace test
     }
 
 #if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_UNORDERED_NOEXCEPT
+    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_NOEXCEPT
     {
       cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
       return *this;

--- a/test/objects/cxx11_allocator.hpp
+++ b/test/objects/cxx11_allocator.hpp
@@ -179,17 +179,6 @@ namespace test
       detail::tracker.allocator_ref();
     }
 
-#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator_base(cxx11_allocator_base&& x) BOOST_NOEXCEPT
-    {
-      tag_ = x.tag_;
-      selected_ = x.selected_;
-
-      x.tag_ = -1;
-      x.selected_ = -1;
-    }
-#endif
-
     ~cxx11_allocator_base() { detail::tracker.allocator_unref(); }
 
     cxx11_allocator_base& operator=(cxx11_allocator_base const& x)
@@ -198,17 +187,6 @@ namespace test
       selected_ = x.selected_;
       return *this;
     }
-
-#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator_base& operator=(
-      cxx11_allocator_base&& x) BOOST_NOEXCEPT
-    {
-      tag_ = x.tag_;
-      selected_ = x.selected_;
-
-      return *this;
-    }
-#endif
 
     pointer address(reference r) { return pointer(&r); }
 
@@ -300,14 +278,6 @@ namespace test
       return *this;
     }
 
-#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_NOEXCEPT
-    {
-      cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
-      return *this;
-    }
-#endif
-
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.
     bool operator==(cxx11_allocator const& x) const
@@ -356,14 +326,6 @@ namespace test
       cxx11_allocator_base<T>::operator=(x);
       return *this;
     }
-
-#if !defined(BOOST_NO_CXX11_RVALUE_REFERENCES)
-    cxx11_allocator& operator=(cxx11_allocator&& x) BOOST_NOEXCEPT
-    {
-      cxx11_allocator_base<T>::operator=(static_cast<cxx11_allocator&&>(x));
-      return *this;
-    }
-#endif
 
     // When not propagating swap, allocators are always equal
     // to avoid undefined behaviour.

